File: 2015/bryan/day05.py
# Time: O(N)
# Space: O(1)


def check_invalid(s):
    for i in range(len(s)-1):
        if s[i:i+2] in ['ab', 'cd', 'pq', 'xy']:
            # print(f"string {s} is invalid")
            return True
    return False

def check_twice(s):
    for i in range(len(s)-1):
        if s[i] == s[i+1]:
            # print(f"string {s} has two letters in a row")
            return True
    return False

def count_vowels(s):
    vowel_count = 0
    for ch in s:
        if ch in 'aeiou':
            vowel_count += 1
    # print(f"string {s} has {vowel_count} vowels")
    return vowel_count

def is_nice(s):
    if (count_vowels(s) >= 3) and check_twice(s) and not check_invalid(s):
        return True
    else:
        return False

def count_nice(fpath):
    num_nice = 0
    num_naughty = 0
    with open(fpath, 'r') as f:
        for line in f:
            if is_nice(line):
                num_nice += 1
            else:
                num_naughty += 1
    return num_nice

assert is_nice('ugknbfddgicrmopn')
assert is_nice('aaa')
assert not is_nice('jchzalrnumimnmhp')
assert not is_nice('haegwjzuvuyypxyu')
assert not is_nice('dvszwmarrgswjxmb')

print(count_nice('Inputs/day05.txt'))

def has_twice_without_overlaps(s):
    pairs_to_positions = {}
    for i in range(len(s)-1):
        pair = s[i:i+2]
        if pair in pairs_to_positions:
            if i in pairs_to_positions[pair]:
                continue
            return True
        pairs_to_positions[pair] = set([i, i+1])
    return False

def has_repeat_with_middle(s):
    for i in range(len(s)-2):
        if s[i] == s[i+2]:
            return True
    return False

def is_nice2(s):
    if has_repeat_with_middle(s) and has_twice_without_overlaps(s):
        return True
    else:
        return False

def count_nice2(fpath):
    num_nice = 0
    with open(fpath, 'r') as f:
        for line in f:
            if is_nice2(line):
                num_nice += 1

    return num_nice

assert is_nice2('qjhvhtzxzqqjkmpb')
assert is_nice2('xxyxx')
assert not is_nice2('uurcxstgmygtbstg')
assert not is_nice2('ieodomkazucvgmuy')

print(count_nice2('Inputs/day05.txt'))

File: 2015/nate/day05/day05.js
// A nice string is one with all of the following properties:

// It contains at least three vowels (aeiou only), like aei, xazegov, or aeiouaeiouaeiou.
// It contains at least one letter that appears twice in a row, like xx, abcdde (dd), or aabbccdd (aa, bb, cc, or dd).
// It does not contain the strings ab, cd, pq, or xy, even if they are part of one of the other requirements.
// For example:

// ugknbfddgicrmopn is nice because it has at least three vowels (u...i...o...), a double letter (...dd...), and none of the disallowed substrings.
// aaa is nice because it has at least three vowels and a double letter, even though the letters used by different rules overlap.
// jchzalrnumimnmhp is naughty because it has no double letter.
// haegwjzuvuyypxyu is naughty because it contains the string xy.
// dvszwmarrgswjxmb is naughty because it contains only one vowel.

//ugknbfddgicrmopn === nice
//aaa === nice
//jchzalrnumimnmhp === naughty
//haegwjzuvuyypxyu === naughty
//dvszwmarrgswjxmb === naughty

const fs = require('fs')
const path = require('path')

function check3Vowels(string) {
    let has3Vowels = false

    const vowels = string.match(/[aeiou]/g)

    if (vowels !== null && vowels.length >= 3) {
        has3Vowels = true
    }
    return has3Vowels
}

function checkDoubleLet(string) {
    let hasDouble = false

    string.split('').forEach((letter, index) => {

        if (index < (string.length - 1) && (letter === string.charAt(index + 1))) {
            hasDouble = true
        }

    });

    return hasDouble
}

function forbbidenSubStr(string) {
    let containsForbidden = false

    const forbidden = ['ab', 'cd', 'pq', 'xy']

    forbidden.forEach(item => {

        if (string.includes(item)) {
            containsForbidden = true
        }

    })

    return containsForbidden
}


function checkNice(string) {
    return check3Vowels(string) && checkDoubleLet(string) && !forbbidenSubStr(string)
}

// console.log(
//     '//ugknbfddgicrmopn === nice',
//     checkNice('ugknbfddgicrmopn')
// )

// console.log(
//     '//aaa === nice',
//     checkNice('aaa')
// )

// console.log(
//     'jchzalrnumimnmhp === naughty',
//     checkNice('jchzalrnumimnmhp')
// )

// console.log(
//     'haegwjzuvuyypxyu === naughty',
//     checkNice('haegwjzuvuyypxyu')
// )

// console.log(
//     'dvszwmarrgswjxmb === naughty',
//     checkNice('dvszwmarrgswjxmb')
// )

// function checkNice2(string) {
//     return
// }

// console.log(
//     'jchzalrnumimnmhp === naughty',
//     checkNice('jchzalrnumimnmhp')
// )

// console.log(
//     'haegwjzuvuyypxyu === naughty',
//     checkNice('haegwjzuvuyypxyu')
// )

// console.log(
//     'dvszwmarrgswjxmb === naughty',
//     checkNice('dvszwmarrgswjxmb')
// )

// Part 2

// 1. It contains a pair of any two letters that appears at least twice in the string without overlapping, like xyxy (xy) or aabcdefgaa (aa), but not like aaa (aa, but it overlaps).

// 2. It contains at least one letter which repeats with exactly one letter between them, like xyx, abcdefeghi (efe), or even aaa.

function checkNice2(string) {
    return checkDoubleNoOverlap(string) && checkDoubleLetterBetween(string)
}

function checkDoubleNoOverlap(string){
    for (let i = 0 ; i < string.length - 1; i++){
        const pair = string.charAt(i) + string.charAt(i+1)

        for (let j = i+2; j < string.length - 1; j++) {
            const followingPair = string.charAt(j) + string.charAt(j+1)

            // console.log(pair, followingPair)

            if(pair === followingPair){
                return true
            }
        }
    }
    return false
}

function checkDoubleLetterBetween(string){
    let hasDouble = false

    string.split('').forEach((letter, index) => {

        if (index < (string.length - 1) && (letter === string.charAt(index + 2))) {
            hasDouble = true
        }

    });

    return hasDouble

}

//qjhvhtzxzqqjkmpb is nice because is has a pair that appears twice (qj) and a letter that repeats with exactly one letter between them (zxz).
// xxyxx is nice because it has a pair that appears twice and a letter that repeats with one between, even though the letters used by each rule overlap.
// uurcxstgmygtbstg is naughty because it has a pair (tg) but no repeat with a single letter between them.
// ieodomkazucvgmuy is naughty because it has a repeating letter with one between (odo), but no pair that appears twice.

console.log(
    'qjhvhtzxzqqjkmpb === nice',
    checkNice2('qjhvhtzxzqqjkmpb')
)

console.log(
    'xxyxx === nice',
    checkNice2('xxyxx')
)

console.log(
    'uurcxstgmygtbstg === naughty',
    checkNice2('uurcxstgmygtbstg')
)

console.log(
    'ieodomkazucvgmuy === naughty',
    checkNice2('ieodomkazucvgmuy')
)


function solve() {
    const input = fs.readFileSync("../Inputs/day05.txt", 'utf8')

    const inputArr = input.split('\n')
    const niceWords = inputArr.filter(string => checkNice(string))
    const niceWords2 = inputArr.filter(string => checkNice2(string))

    console.log("Part 1, # of nice strings: ", niceWords.length)
    console.log("Part 2, # of nice strings: ", niceWords2.length)

}

solve()
